packaging/metadata.py

```diff
@@ -0,0 +1,191 @@
+        download_url: Optional[str] = None,
+        classifiers: Optional[Iterable[str]] = None,  # TODO: OK?
+        # 1.2
+        maintainer: Optional[str] = None,
+        maintainer_emails: Optional[Iterable[_NameAndEmail]] = None,
+        requires_dists: Optional[Iterable[requirements.Requirement]] = None,
+        requires_python: Optional[specifiers.SpecifierSet] = None,
+        requires_externals: Optional[Iterable[str]] = None,  # TODO: OK?
+        project_urls: Optional[Iterable[_LabelAndURL]] = None,
+        provides_dists: Optional[Iterable[requirements.Requirement]] = None,
+        obsoletes_dists: Optional[Iterable[requirements.Requirement]] = None,
+        # 2.1
+        description_content_type: Optional[str] = None,  # TODO: OK?
+        provides_extras: Optional[Iterable[utils.NormalizedName]] = None,
+        # 2.2
+        dynamic: Optional[Iterable[DynamicField]] = None,
+    ) -> None:
+        """
+        Set all attributes on the instance.
+
+        An argument of `None` will be converted to an appropriate, false-y value
+        (e.g. the empty string).
+        """
+        self.display_name = name
+        self.version = version
+        self.platforms = set(platforms or [])
+        self.summary = summary or ""
+        self.description = description or ""
+        self.keywords = list(keywords or [])
+        self.home_page = home_page or ""
+        self.author = author or ""
+        self.author_emails = list(author_emails or [])
+        self.license = license or ""
+        self.supported_platforms = set(supported_platforms or [])
+        self.download_url = download_url or ""
+        self.classifiers = set(classifiers or [])
+        self.maintainer = maintainer or ""
+        self.maintainer_emails = list(maintainer_emails or [])
+        self.requires_dists = set(requires_dists or [])
+        self.requires_python = requires_python or specifiers.SpecifierSet()
+        self.requires_externals = set(requires_externals or [])
+        self.project_urls = set(project_urls or [])
+        self.provides_dists = set(provides_dists or [])
+        self.obsoletes_dists = set(obsoletes_dists or [])
+        self.description_content_type = description_content_type or ""
+        self.provides_extras = set(provides_extras or [])
+        self.dynamic = set(dynamic or [])
+
+    @property
+    def display_name(self) -> str:
+        return self._display_name
+
+    @display_name.setter
+    def display_name(self, value, /) -> None:
+        """Set the value for self.display_name and self.canonical_name."""
+        self._display_name = value
+        self._canonical_name = utils.canonicalize_name(value)
+
+    # Use functools.cached_property once Python 3.7 support is dropped.
+    # Value is set by self.display_name.setter to keep in sync with self.display_name.
+    @property
+    def canonical_name(self) -> utils.NormalizedName:
+        return self._canonical_name
```

Comment:
I have code (in https://github.com/pfmoore/pkg_metadata) that converts metadata from "email message" form to the "JSON-compatible dictionary" format defined in [PEP 566](https://www.python.org/dev/peps/pep-0566). Would that be of interest here? We could have `from_dict` and `from_msg` classmethods, and `as_dict` and `as_msg` methods.

That library also has code to read `pyproject.toml` PEP 621 data, and create a metadata dictionary from it. Again, is that of interest here?

I'd like to help move this forward, as I have a fair amount of code that would benefit from not repeatedly rewriting this functionality 😉

Reply:
I want to eventually have:

- `from_pyproject()`/`from_toml()`
- `from_email_headers()`
- `to_email_headers()`
- `from_json()`
- `to_json()`

In all of those cases I'm assuming some `bytes` input/output as I view this class as the typical dict intermediary format that's dealing with stuff read from some wheel or sdist. But I'm also totally open to discussing the idea of passing around dicts as well.

My motivation for thinking `bytes` is it lets the class control for encoding more so that people don't mess up and don't write UTF-8. But maybe that's an unnecessary concern?

Reply:
If you have `from_json` and `to_json`[^1], the rest are easy to build based on the functions in `pgk_metadata`.

Strong -1 on considering this class as the intermediary format. Rejecting invalid data and lack of roundtripping kills that for me. For me, the JSON-compatible dictionary form is the intermediary, because it handles the raw data *without* interpreting it. The class is the "programmer API" for metadata, which adds a convenience layer over the raw data.

So I'd have `from_dict` and `to_dict`, with all of the rest being standalone functions that produce or consume the dict format.

In my experience:
* The JSON methods should accept and return JSON-compatible dictionaries, *not* actual JSON data (strings or bytes). I'm not entirely sure if this is what you meant, but that's what I'm assuming.
* The pyproject method should take a dictionary. The caller has probably already read the TOML file by this point, and will have the data in dict form.
* I agree that the email writer should produce a bytes object. I got that wrong in `pkg_metadata`, returning a string. I need to fix that - now that UTF-8 is part of the spec, we should enforce it.
* The email reader could take bytes. But older versions of the spec didn't mandate UTF-8, and older tools do seem to have used the platform encoding. So it's technically wrong to fail on non-UTF8 input (and in practical terms, it's unhelpful to reject a whole metadata file just because there's one undecodable byte in a field like `Description` that you're not planning on using). My implementation attempts to use UTF-8 and falls back to latin1 (risking mojibake rather than failing) on a per-field basis. I guess we could have a "strict" parameter that fails if the input isn't valid UTF-8, but I know I'd pretty much always set it to false (even in pip). So I'd prefer to just be non-strict on input. What I'd be very much against is any form of processing that doesn't decide what to do per-field. There are far too many use cases for reading data that uses only ASCII for the fields you care about, so ignoring encoding problems in other fields is critical.

[^1]: I prefer `from_dict` and `to_dict`, as I don't think we should be serialising to bytes here, just return the dictionary data. It's a more useful format, and everyone knows how to serialise a dictionary to JSON.

Reply:
> Strong -1 on considering this class as the intermediary format. Rejecting invalid data and lack of roundtripping kills that for me. For me, the JSON-compatible dictionary form is the intermediary, because it handles the raw data _without_ interpreting it. The class is the "programmer API" for metadata, which adds a convenience layer over the raw data.

But I would also like to avoid having the extra data translation step of JSON dict to structured class. See my proposal above on a compromise class structure.